dohproxy/server_protocol.py

```diff
@@ -60,11 +118,29 @@ def datagram_received(self, data, addr):
                 interval
             )
         )
-        self.queue.put_nowait(dnsr)
+        self.fut.set_result(dnsr)
         self.transport.close()
 
-    def error_received(self, exc):
-        print('Error received:', exc)
 
-    def connection_lost(self, exc):
-        pass
+class DNSClientProtocolUDP(DNSClientProtocol):
+
+    def connection_made(self, transport):
+        self.send_helper(transport)
+        self.transport.sendto(self.dnsq.to_wire())
+
+    def datagram_received(self, data, addr):
+        dnsr = dns.message.from_wire(data)
+        self.receive_helper(dnsr)
+
+
+class DNSClientProtocolTCP(DNSClientProtocol):
+
+    def connection_made(self, transport):
+        self.send_helper(transport)
+        msg = self.dnsq.to_wire()
+        tcpmsg = struct.pack("!H", len(msg)) + msg
+        self.transport.write(tcpmsg)
+
+    def data_received(self, data):
```

Comment:
this may be called multiple time and everytime it is called, you should happen the content to what you have already received. As is, it will work because messages are small enough . But it would most likely fail for somethig like: `dig axfr . @b.root-servers.net`
https://docs.python.org/3/library/asyncio-protocol.html#streaming-protocols
Eventually, you will get `eof_received` which will signal that no more data will be sent and then you will be able to get the whole message, build the DNS response from it and sent it back using the `receive_helper`.

This will work under the assumption that the connection will not be re=used. Otherwise, you should read the 2 first bytes you receive and then know you will expect `msglen = struct.unpack("!H", data[0:2])[0]` bytes after.
You will keep adding to an internal buffer until `msglen` bytes are available, and then you can generate thee dns response.